### PR TITLE
Make sans idf sensitive and update larmor idf

### LINKED ISSUE #13682: Update the IDF for Larmor
sans_feature
Leo wrote:

>I’ve attached the Larmor definition file (updated 10June2015) that needs to be included in future Mantid builds. This definition file has updated tube positions. The installation of our new detector meant that the tubes are now closer together.

### PR DESCRIPTION
**Pull request by AntonPiccardoSelg**
Fixes #13682 

This ticket provides new IDFs for LARMOR and enables the automatic switching of the IDF which stored in the isis_instrument.

Until now the IDF was not chosen automatically, this meant that no further  IDFs could be added, or at least the correct ones would not be picked up.

Implementation is tricky, since we first create the isis_instrument with an associated IDF and then load a workspace. At the point when we load the worksapce we compare the IDF of the isis_instrument with the IDF of the workspace. If they don't match the IDf of the isis_instrument is updated. The user files has to be reloaded at this point as well (annoying but the current architecture requires this).
# For Testing

For each IDF version of LOQ, SANS2D and LARMOR a sample reduction needs to be executed on a Mantid version which contains these changes. The system tests still pass. Since there are three instruments (LOQ, SANS2D and LARMOR) and several IDF versions, there are quite a few tests to execute.

LOQ has 2 IDFs
SANS2D has 2 IDFs
LARMOR has 3IDFs
## LARMOR

**IDF1**

Note that there are several warnings which come up for `base instrument` and `geometry`. They are unrelated to the changes here.
1. Setup that Mantid points to the folder LARMOR/IDF1
2. Open the SANS GUI and select LARMOR
3. Load the user file USER_LARMORTEAM_152L_a2_6x8mm_r5889.txt
   - Confirm that the field `Current IDF` shows the path to  LARMOR_Definition.xml
4. Load the LARMOR00000063.nxs
   - Confirm that the field `Current IDF` shows the path to  LARMOR_Definition_19000000-20150317.xml
5. Run `1D Reduce`
   - Confirm that there are no errors

**IDF2**

Note that there are several warnings which come up for `base instrument`. They are unrelated to the changes here.
1. Setup that Mantid points to the folder LARMOR/IDF2
2. Open the SANS GUI and select LARMOR
3. Load the user file USER_LARMORTEAM_152L_a2_6x8mm_r5889.txt
   - Confirm that the field `Current IDF` shows the path to  LARMOR_Definition.xml
4. Load the LARMOR00002260.nxs
   - Confirm that the field `Current IDF` shows the path to  LARMOR_Definition_8tubes.xml

We cannot run the reduction here, since the data file is not a very good LARMOR file (I dont have any other currently). The important bit is the change in the IDF

**IDF3**

Note that there are several warnings which come up for `base instrument`. They are unrelated to the changes here.
1. Setup that Mantid points to the folder LARMOR/IDF3
2. Open the SANS GUI and select LARMOR
3. Load the user file USER_LARMORTEAM_152L_a2_6x8mm_r5889.txt
   - Confirm that the field `Current IDF` shows the path to  LARMOR_Definition.xml
4. Load the LARMOR00004735.nxs
   - Confirm that the field `Current IDF` shows the path to  LARMOR_Definition.xml
5. Run `1D Reduce`
   - Confirm that there are no errors
## SANS2D

**SANS2D**
1. Setup that Mantid points to the folder SANS2D/SANS2D
2. Open the SANS GUI and select SANS2D (!!!! this actually sets an IDF)
3. Load the user file MaskSANS2DReductionGUI.txt
   - Confirm that the field `Current IDF` shows the path to  SANS2D_Definition.xml
4. Load the following:
   - Sample    22048    22041  22024
   - Can         22023    22024  22024
   - Confirm that the field `Current IDF` shows the path to  SANS2D_Definition.xml
5. Run `1D Reduce`
   - Confirm that there are no errors

**SANS2DTUBES**

There will be a warning about values differing and spectra unaffected, but this is not related to the changes here.
1. Setup that Mantid points to the folder SANS2D/SANS2DTUBES
2. Open the SANS GUI and select SANS2DTUBES (!!!! this actually sets an IDF)
3. Load the user file USER_SANS2D_143ZC_2p4_4m_M4_Knowles_12mm.txt
   - Confirm that the field `Current IDF` shows the path to  SANS2D_Definition_Tubes.xml
4. Load the following:
   - Sample    28797    28808  28824
   - Can         28793    28823  28824
   - Confirm that the field `Current IDF` shows the path to  SANS2D_Definition_Tubes.xml
5. Run `1D Reduce`
   - Confirm that there are no errors
## LOQ

**IDF1**

Note that a lot of warnings appear here. We probably don't use the correct user file for this.
1. Setup that Mantid points to the folder LOQ/IDF1
2. Open the SANS GUI and select LOQ
3. Load the user file MASK.094AA
   - Confirm that the field `Current IDF` shows the path to LOQ_Definition_20121016-.xml
4. Load the following:
   - Sample    54431  54435  54433
   - Can         54432  54434  54433
   - Confirm that the field `Current IDF` shows the path to  LOQ_Definition_20020226-.xml
5. Run `1D Reduce`
   - Confirm that there are no errors

**IDF2**
1. Setup that Mantid points to the folder LOQ/IDF1
2. Open the SANS GUI and select LOQ
3. Load the user file USER_LOQ_143A_TeamLOQ.txt
   - Confirm that the field `Current IDF` shows the path to LOQ_Definition_20121016-.xml
4. Load the following:
   - Sample    LOQ94458
   - Confirm that the field `Current IDF` shows the path to   LOQ_Definition_20121016-.xml
5. Run `1D Reduce`
   - Confirm that there are no errors

### CODE REVIEW

File: Testing/SystemTests/tests/analysis/SANSFileChecking.py [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+        return idf_path_workspace
+
+    def test_that_reducer_for_SANS2D_switches_to_correct_IDF_when_outdated(self):
+        # Arrange
+        Clean()
+        SANS2D()
+        MaskFile('MASKSANS2D.091A')
+        Set1D()
+        instrument_name = "SANS2D"
+        #filename = 'SANS2D00000992.raw'
```
Comment (matthew-d-jones) [inline Testing/SystemTests/tests/analysis/SANSFileChecking.py]: Can this commented out filename be removed? (also line 120)

File: scripts/SANS/isis_reducer.py [hunk 2]
```diff
@@ -716,6 +731,84 @@ def update_beam_center(self):
...
+        # Get the idf from the reducer
+        idf_path_reducer = self.get_idf_file_path()
+        idf_path_reducer = os.path.normpath(idf_path_reducer)
+
+        # Now check if both idf paths and underlying files. If they are, then don't do anything
+        # else switch the underlying instrument
+        if idf_path_reducer == idf_path_workspace and su.are_two_files_identical(idf_path_reducer, idf_path_reducer):
+            return
+        else:
+            logger.warning("Updating the IDF of the Reducer. Switching from " + str(idf_path_reducer) + " to " + str(idf_path_workspace))
```
Comment (matthew-d-jones) [inline scripts/SANS/isis_reducer.py]: This is a really long line...

File: scripts/SANS/isis_reducer.py [hunk 3]
```diff
@@ -716,6 +731,84 @@ def update_beam_center(self):
...
+        user file.
+        @param run: name of the run for which the file is to be extracted
+        '''
+        # We need the instrument name and the measurement time to determine
+        # the IDF
+        measurement_time = None
+        instrument_name = self.get_instrument_name()
+        # We need to be able to handle file-based and workspace-based queries
+        # If we have a workspace we look at the end time, else we
+        # need a sophisticated extraciton mechanism
```
Comment (matthew-d-jones) [inline scripts/SANS/isis_reducer.py]: Typo in comment

File: scripts/SANS/isis_reducer.py [hunk 4]
```diff
@@ -716,6 +731,84 @@ def update_beam_center(self):
...
+
+            if instrument is not None:
+                self.set_instrument(instrument)
+
+                # We need to update the instrument, by reloading the user file.
+                # This is pretty bad, but looking at the reducer architecture this
+                # seems to be the only reasonable way to do this.
+                self.user_settings.execute(self)
+
+                # Now we set the corret detector, this is also being done in the GUI
```
Comment (matthew-d-jones) [inline scripts/SANS/isis_reducer.py]: Typo in comment

File: scripts/SANS/isis_reducer.py (hunk 2 above)
Comment (picatess) [inline scripts/SANS/isis_reducer.py]: When it is displayed in the logs or in the code?

Comment (matthew-d-jones) [inline scripts/SANS/isis_reducer.py]: In the code

**Comment by matthew-d-jones** [comment]
Release notes were updated. Code looks good and the described testing had the expected results.
:shipit:

**Pull request merged**